model/speedscope_profile.go

```diff
@@ -38,41 +40,373 @@ type SpeedscopeProfileEvent struct {
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		return err
+	}
+
+	val, ok := objMap["$schema"]
+	if !ok {
+		return fmt.Errorf("field $schema is required")
+	}
+	var schema string
+	err = json.Unmarshal(*val, &schema)
+	if err != nil {
+		return err
+	}
+	p.Schema = &schema
+
+	val, ok = objMap["shared"]
+	if !ok {
+		return fmt.Errorf("field shared is required")
+	}
+	var shared Shared
+	err = json.Unmarshal(*val, &shared)
+	if err != nil {
+		return err
+	}
+	p.Shared = shared
+
+	// handle profiles union type
+	val, ok = objMap["profiles"]
+	if !ok {
+		return fmt.Errorf("field profiles is required")
+	}
+	var profiles []*json.RawMessage
+	err = json.Unmarshal(*val, &profiles)
+	if err != nil {
+		return err
+	}
+
+	p.Profiles = make([]Profile, len(profiles))
+	var raw map[string]interface{}
+	for index, rawMessage := range profiles {
+		err = json.Unmarshal(*rawMessage, &raw)
+		if err != nil {
+			return err
+		}
+		if raw["type"] == "evented" {
+			var e EventedProfile
+			err := json.Unmarshal(*rawMessage, &e)
+			if err != nil {
+				return err
+			}
+			p.Profiles[index] = &e
+		} else if raw["type"] == "sampled" {
+			var s SampledProfile
+			err := json.Unmarshal(*rawMessage, &s)
+			if err != nil {
+				return err
+			}
+			p.Profiles[index] = &s
+		} else {
+			return fmt.Errorf("unsupported type %#v", raw["type"])
+		}
+	}
+
+	if val, ok := objMap["name"]; ok {
+		var name string
+		err = json.Unmarshal(*val, &name)
+		if err != nil {
+			return err
+		}
+		p.Name = &name
+	}
+
+	if val, ok := objMap["exporter"]; ok {
+		var exporter string
+		err = json.Unmarshal(*val, &exporter)
+		if err != nil {
+			return err
+		}
+		p.Exporter = &exporter
+	}
+
+	if val, ok := objMap["activeProfileIndex"]; ok {
+		var activeProfileIndex float64
+		err = json.Unmarshal(*val, &activeProfileIndex)
+		if err != nil {
+			return err
+		}
+		p.ActiveProfileIndex = &activeProfileIndex
+	}
+
+	return nil
+}
+
+func (e *SampledProfile) UnmarshalJSON(b []byte) error {
+	var raw map[string]interface{}
+	if err := json.Unmarshal(b, &raw); err != nil {
+		return err
+	}
+	if v, ok := raw["endValue"]; !ok || v == nil {
+		return fmt.Errorf("field endValue: required")
+	}
+	if v, ok := raw["name"]; !ok || v == nil {
+		return fmt.Errorf("field name: required")
+	}
+	if v, ok := raw["samples"]; !ok || v == nil {
+		return fmt.Errorf("field samples: required")
+	}
+	if v, ok := raw["startValue"]; !ok || v == nil {
+		return fmt.Errorf("field startValue: required")
+	}
+	if v, ok := raw["type"]; !ok || v == nil {
+		return fmt.Errorf("field type: required")
+	}
+	if v, ok := raw["unit"]; !ok || v == nil {
+		return fmt.Errorf("field unit: required")
+	}
+	if v, ok := raw["weights"]; !ok || v == nil {
+		return fmt.Errorf("field weights: required")
+	}
+	type Plain SampledProfile
+	var plain Plain
+	if err := json.Unmarshal(b, &plain); err != nil {
+		return err
+	}
+	*e = SampledProfile(plain)
```

Comment:
I would not have come up with this. :)  This isn't criticism -- I don't understand why the `type Plain` dance is necessary. <shrug for now>

I think all this parsing would eventually benefit from one of the apm-server folks reviewing.

Reply:
I _think_ the `Plain` type aliasing is to prevent a stackoverflow; the type aliasing avoids `json.Unmarshal()` from calling into the deserialization of the type in question (in this case `SampledProfile`) recursively. In effect, the `UnmarshalJSON()` implementation for a speedscope type is unmarshal to validate, then type alias to perform the actual deserialization, if valid.